keps/sig-autoscaling/20190307-configurable-scale-velocity-for-hpa.md

```diff
@@ -0,0 +1,444 @@
+[Algorithm Details]: https://v1-14.docs.kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#algorithm-details
+
+#### API Changes
+
+The following API changes are needed:
+
+We should add `Scale{Up,Down}{Percent,Pods}` fields into the HPA spec
+
+The resulting solution will look like this:
+
+```golang
+type HPAScaleConstraintValue struct {
+    Rate         *HPAScaleConstraintRateValue
+    DelaySeconds *int32
+
+type HPAScaleConstraintRateValue struct {
+    Pods          *int32
+    Percent       *int32
+    PeriodSeconds *int32
+}
+
+type HPAScaleConstraints struct {
+    ScaleUp   *HPAScaleConstraintValue
+    ScaleDown *HPAScaleConstraintValue
+}
+
+type HorizontalPodAutoscalerSpec struct {
+    ScaleTargetRef CrossVersionObjectReference
+    MinReplicas    *int32
+    MaxReplicas    int32
+    Metrics        []MetricSpec
+    Constraints    *HPAScaleConstraints
+}
+```
+
+#### HPA Controller State Changes
+
+To store not only scale down recommendations, we need to replace
+
+```golang
+    recommendations map[string][]timestampedRecommendation
+```
+
+with
+
+```golang
+    scaleUpRecommendations   map[string][]timestampedRecommendation
+    scaleDownRecommendations map[string][]timestampedRecommendation
+```
+
+To store the information about last scale action, the HPA need an additional field (similar to the list of “recommendations”)
+
+```golang
+scaleEvents map[string][]timestampedScaleEvent
+```
+
+Where
+
+```golang
+type timestampedScaleEvent struct {
+    replicaChange int32
+    timestamp     time.Time
+}
+```
+
+It will store last scale events, and it will be used to make decisions about next scale events.
+
+Say, if 30 seconds ago the number of replicas was increased by one, and we forbid to scale up for more than 1 pod per minute,
+then during the next 30 seconds, we won’t scale up again.
+
+If the HPA is restarted, the state information is lost so that it might scale the cluster instantly after the restart.
+Though, I don’t think this is a problem, as:
+
+- It shouldn’t happen often, or you have some cluster problem
+- It requires quite a lot of time to start an HPA pod, for HPA pod to become live and ready, to get and process metrics.
+- If you have a large discrepancy between what is a desired number of replicas according to metrics and what is your current number of replicas and you DON’T want to scale - probably, you shouldn’t want to use the HPA. As the HPA goal is the opposite.
+
+As the added parameters have default values, we don’t need to update the API version, and may stay on the same `pkg/apis/autoscaling/v2beta2`.
+
+#### Command Line Options Changes
+
+First, we should note that the
+current [--horizontal-pod-autoscaler-downscale-stabilization-window][] option
+defines the default value for the `constraints.scaleDown.delaySeconds`
+So, we should make it obsolete and warn users to switch to a new command line option described below.
+
+We should implement a new command line options
```

Comment:
I don't think we should change the command line options.  We should support `--horizontal-pod-autoscaler-downscale-stabilization-window` for backward compatibility.  But command line options is not a good way to configure defaults.  ConfigMaps for example would be better.

Reply:
Ok, I removed these new options implementation: https://github.com/kubernetes/enhancements/pull/883/commits/20831756ecf3fa03115d4080131472c83bfda835